# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFrameLoaderOwner.h
# Commit: d027f301f08c
# Full Hash: d027f301f08c38d8ab3095abfe0a04f2255f6e6c
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-06-16 03:30:48
# Regressor Bug: 1640019
# File Overlap Count: 1
# Description:
#   Bug 1640019 - Part 1: Support toplevel process switches outside of tabbrowser, r=mattwoodrow
#   
#   This new process switching behavior is only enabled for some browser elements,
#   which have specified a specific attribute. Turning this on for all browsers with
#   a `remote` attribute causes breakage in reftests.
# ==============================================================================

diff -r 82becb1f6eae -r d027f301f08c dom/base/nsFrameLoaderOwner.h
--- a/dom/base/nsFrameLoaderOwner.h	Tue Jun 16 02:46:15 2020 +0300
+++ b/dom/base/nsFrameLoaderOwner.h	Mon Jun 15 23:23:43 2020 +0000
@@ -46,23 +46,37 @@
   mozilla::dom::BrowsingContext* GetExtantBrowsingContext();
 
   // Destroy (if it exists) and recreate our frameloader, based on new
-  // remoteness requirements. This should follow the same path as
-  // tabbrowser.js's updateBrowserRemoteness, including running the same logic
-  // and firing the same events as unbinding a XULBrowserElement from the tree.
-  // However, this method is available from backend and does not manipulate the
-  // DOM.
+  // remoteness requirements.
+  //
+  // This method is called by frontend code when it wants to perform a
+  // remoteness update, and allows for behaviour such as preserving
+  // BrowsingContexts across process switches during navigation.
+  //
+  // See the WebIDL definition for more details.
   void ChangeRemoteness(const mozilla::dom::RemotenessOptions& aOptions,
                         mozilla::ErrorResult& rv);
 
+  // Like `ChangeRemoteness` but switches to an already-created
+  // `BrowserBridgeChild`. This method is used when performing remote subframe
+  // process switches.
   void ChangeRemotenessWithBridge(mozilla::dom::BrowserBridgeChild* aBridge,
                                   mozilla::ErrorResult& rv);
 
+  // Like `ChangeRemoteness`, but switches into an already-created
+  // `ContentParent`. This method is used when performing toplevel process
+  // switches. If `aContentParent` is nullptr, switches into the parent process.
+  //
+  // If `aReplaceBrowsingContext` is set, BrowsingContext preservation will be
+  // disabled for this process switch.
+  void ChangeRemotenessToProcess(mozilla::dom::ContentParent* aContentParent,
+                                 uint64_t aPendingSwitchId,
+                                 bool aReplaceBrowsingContext,
+                                 mozilla::ErrorResult& rv);
+
   void SubframeCrashed();
 
  private:
   bool UseRemoteSubframes();
-  bool ShouldPreserveBrowsingContext(
-      const mozilla::dom::RemotenessOptions& aOptions);
 
   // The enum class for determine how to handle previous BrowsingContext during
   // the change remoteness. It could be followings
@@ -78,9 +92,11 @@
     DONT_PRESERVE_BUT_PROPAGATE = 1,
     PRESERVE = 2,
   };
+  ChangeRemotenessContextType ShouldPreserveBrowsingContext(
+      bool aIsRemote, bool aReplaceBrowsingContext);
+
   void ChangeRemotenessCommon(const ChangeRemotenessContextType& aContextType,
-                              bool aSwitchingInProgressLoad,
-                              const nsAString& aRemoteType,
+                              bool aSwitchingInProgressLoad, bool aIsRemote,
                               std::function<void()>& aFrameLoaderInit,
                               mozilla::ErrorResult& aRv);
 